<!DOCTYPE html>
<html>
<head>
<title>Sequencing_pipeline.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h1 id="course-protocol">Course protocol</h1>
<h1 id="sequencing-pipeline">Sequencing Pipeline</h1>
<p>To go from raw data produced by a sequencer to a completed genome many steps are needed. These steps can be summarized under the term &quot;sequencing pipeline&quot; and are described below. In each step, the data is processed and the output is used as input for the next step.
By utilizing batch scripts to run the pipeline on a cluster, it is possible to run commands on multiple samples at the same time. In addition, the resources provided by the cluster decrease the time it takes to run the scripts significantly.</p>
<h2 id="cluster">Cluster</h2>
<p>The cluster is a high performance computer linked to the internet, it is used to run computationally intensive programs. To use the cluster, you need to connect to it using a terminal. Then you can run commands on the cluster, for example Bash scripts. The commands are sent to the cluster and the cluster runs the commands. The results are then sent back to your computer.</p>
<p>Important commands:</p>
<ul>
<li><code>ssh -X username@clustername</code> - Connect to the cluster</li>
<li><code>sbatch batch_name</code> - Run a Batch script on the cluster</li>
<li><code>squeue -u username</code> - Check the status of the jobs on the cluster</li>
<li><code>scancel job_id</code> - Cancel a job on the cluster</li>
<li><code>sinfo</code> - Check the status of the cluster</li>
<li><code>exit</code> - Disconnect from the cluster</li>
</ul>
<h2 id="batch-scripts">Batch scripts</h2>
<p>Batch scripts are Bash scripts that are run on the cluster. The scripts are used to run commands on multiple samples at the same time. The Format is <code>script_name.sh</code>. At the beginning of the script, you need to specify the resources you need to run the script, as well as the output name.
An example of a batch script is shown below.</p>
<pre class="hljs"><code><div>#!/bin/bash
#SBATCH -J job_name
#SBATCH -o job_name.out
#SBATCH -e job_name.err
#SBATCH -p short
#SBATCH -n 1
#SBATCH -t 0-00:30
#SBATCH --mem=100
#SBATCH --nodelist=node002

´whatever commands you want to run´
</div></code></pre>
<p>These lines specify:</p>
<ul>
<li>the name of the job</li>
<li>the name of output and error files</li>
<li>the partition</li>
<li>the number of cores</li>
<li>the time limit</li>
<li>the memory limit</li>
<li>the node to run the job on</li>
</ul>
<p>The partition is the queue on the cluster. The time limit is the maximum time the job can run. The memory limit is the maximum amount of memory the job can use.</p>
<p>To run the script, you need to use the command <code>sbatch script_name.sh</code>. The script will then be run on the cluster. The output and error files will be created in the directory where the script is located. The output file contains the output from the script. The error file contains the error messages from the script. The input and output directories for the data can be specified in the script.</p>
<p>There are several batch file extensions for different programming languages:</p>
<ul>
<li><code>.sh</code> - shell script</li>
<li><code>.py</code> - python script</li>
<li><code>.R</code> - R script</li>
</ul>
<p>After submitting the script, you can check the status of the job using the command <code>squeue -u username</code> or cancel the job using the command <code>scancel job_id</code>.</p>
<h2 id="conda-environment">Conda environment</h2>
<p>The sequencing pipeline uses many programs and packages. To make sure that these programs are installed correctly, it is recommended to use a conda environment. A conda environment is a virtual environment that contains all the programs and packages needed to run the pipeline.</p>
<ul>
<li>To create a conda environment, you need to use the command <code>conda create -n environment_name</code>.</li>
<li>To activate the environment, you need to use the command <code>conda activate environment_name</code>.</li>
<li>To install a program in the environment, you need to use the command <code>conda install program_name</code>.</li>
<li>To deactivate the environment, you need to use the command <code>conda deactivate</code>.</li>
</ul>
<p>It might be necessary to install additional modules to run the programs. For example, to run fastqc, you need to install the module <strong>fastqc</strong>. To install a module, you need to use the command <code>module load module_name</code>.</p>
<p>To adjust the parameters of the programs correctly, it is recommended to read the documentation of the programs. The documentation can be found on the internet or by using the command <code>program_name --help</code>.</p>
<p>Look for the following information:</p>
<ul>
<li>input file</li>
<li>output file</li>
<li>essential parameters</li>
<li>optional parameters</li>
<li>required modules</li>
</ul>
<h2 id="sequencing-pipeline">Sequencing pipeline</h2>
<h3 id="overview-of-the-sequencing-pipeline">Overview of the sequencing pipeline</h3>
<ol>
<li>Raw data</li>
<li>Quality control / Trimming</li>
<li>Assembly</li>
<li>Binning</li>
<li>MAGs Quality assessment</li>
<li>Bin refinement</li>
<li>Taxonomic classification</li>
</ol>
<h2 id="1-raw-data">1. Raw data</h2>
<ul>
<li>
<p>Raw data is the output from the sequencer.</p>
</li>
<li>
<p>File format <strong>FASTQ</strong> (example file: SRR2584869_1.fastq.gz)</p>
</li>
<li>
<p>The file contains four lines for each DNA fragment:</p>
<ul>
<li>The first line contains the fragment ID. (optional information like length)</li>
<li>The second line contains the DNA sequence.</li>
<li>A separator, which is simply a plus (+) sign. (optional fragment ID)</li>
<li>The fourth line contains the quality score for each nucleotide in the DNA sequence.</li>
</ul>
<p>The quality score is a number between 0 and 40 encoded in ASCII format. The higher the number, the higher the quality of the nucleotide. The quality score is calculated by the sequencer, it is used to filter out low quality nucleotides.</p>
<p>An example of a FASTQ file is shown below.</p>
<pre class="hljs"><code><div>@SEQ_ID
GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCACAGTTT
+
!''*((((***+))%%%++)(%%%%).1***-+*''))**55CCF&gt;&gt;&gt;&gt;&gt;&gt;CCCCCCC65
</div></code></pre>
</li>
</ul>
<h2 id="2-quality-control--trimming">2. Quality control / Trimming</h2>
<ul>
<li>The quality of the DNA fragments is checked.</li>
<li>The DNA fragments are trimmed to remove low quality nucleotides.</li>
<li>The DNA fragments are filtered to remove fragments with too many low quality nucleotides.</li>
</ul>
<p>The following programs are used:</p>
<ul>
<li>
<p><strong>Fastqc</strong> is a program that checks the quality of the DNA fragments. It creates a report that contains information about the quality of the DNA fragments. The report is used to decide which parameters to use for trimming and filtering.</p>
<ul>
<li>Command: <strong><a href="https://home.cc.umanitoba.ca/~psgendb/doc/fastqc.help">fastqc</a></strong></li>
<li>Input file: <code>fastq</code></li>
<li>Output: report.<code>HTML</code></li>
</ul>
<details>
<summary>Expand to see the documentation of fastqc</summary>
</details><p>fastqc accepts the following parameters:</p>
<ul>
<li><code>-o</code> - output directory</li>
<li><code>-t</code> - number of threads</li>
<li><code>-f</code> - file format</li>
<li><code>-k</code> - kmer size</li>
<li><code>-q</code> - quiet mode</li>
<li><code>-c</code> - number of files to process in parallel</li>
<li><code>-noextract</code> - do not extract the compressed file</li>
<li><code>-nogroup</code> - do not create a summary report</li>
<li><code>-nozip</code> - do not zip the output file</li>
<li><code>-threads</code> - number of threads to use</li>
<li><code>-format</code> - file format</li>
<li><code>-kmers</code> - kmer size</li>
<li><code>-quiet</code> - quiet mode</li>
<li><code>-contaminants</code> - contaminants file</li>
<li><code>-adapters</code> - adapters file</li>
<li><code>-extract</code> - extract the compressed file</li>
<li><code>-n</code> - number of files to process in parallel</li>
<li><code>-s</code> - summary report</li>
<li><code>-z</code> - zip the output file</li>
<li><code>-h</code> - help</li>
<li><code>-v</code> - version</li>
</ul>
<p><a href="https://home.cc.umanitoba.ca/~psgendb/doc/fastqc.help">Full documentation</a></p>
<p>example:</p>
<pre class="hljs"><code><div>#essential (input file is in the current directory):
fastqc input_file

#optional:
fastqc PATH/TO/input_file -o PATH/TO/output_directory -t 4 -f fastq -k 7 -q -c 4 -threads 4 -kmers 7 -n 4  
</div></code></pre>

<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/assemblyscript.sh">Batch script used in course</a></p>
</li>
<li>
<p><strong>Fastp</strong> is a program that trims and filters the DNA fragments. It can be used to trim the DNA fragments to a specific length, to remove low quality nucleotides from the beginning and end of the DNA fragments, and to remove DNA fragments with too many low quality nucleotides. The parameters for trimming and filtering are specified in the batch script. The output of fastp is a fastq file with the trimmed and filtered DNA fragments.</p>
<ul>
<li>Command: <strong><a href="https://github.com/OpenGene/fastp#fastp">fastp</a></strong></li>
<li>Input file: <code>fastq</code> (+ information from report.<code>HTML</code>)</li>
<li>Output file: <code>fastq</code></li>
</ul>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/fastpscript.sh">Batch script used in course</a></p>
</li>
</ul>
<h2 id="3-assembly">3. Assembly</h2>
<ul>
<li>The DNA fragments are assembled into contigs.</li>
<li>Quality assessment of the contigs.
<ul>
<li>Visualization of the contigs.</li>
<li>Analysis of metadata.</li>
</ul>
</li>
</ul>
<p>To assemble the DNA fragments into contigs, the program <strong><a href="https://github.com/voutcn/megahit">megahit</a></strong> is used.</p>
<ul>
<li>
<p>Command: <strong><a href="https://github.com/voutcn/megahit">megahit</a></strong></p>
</li>
<li>
<p>Input: <code>R1_fastq</code> + <code>R2_fastq</code> (trimmed and filtered by fastp)</p>
</li>
<li>
<p>Output: <code>fa</code> (fasta)</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/assemblyscript.sh">Batch script used in course</a></p>
</li>
</ul>
<p>To visualize the contigs, you can use the program <strong><a href="https://rrwick.github.io/Bandage/">Bandage</a></strong>. For this you need to convert the fasta file to a fastg file.</p>
<ul>
<li>
<p>Command: <strong><a href="https://github.com/voutcn/megahit">megahit_toolkit contig2fastg</a></strong></p>
</li>
<li>
<p>Input: <code>fa</code> (fasta)</p>
</li>
<li>
<p>Output: <code>fastg</code></p>
<p>example:</p>
<pre class="hljs"><code><div>megahit_toolkit contig2fastg 99 -i input_file.fa -o output_file.fastg
</div></code></pre>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/contig2fastg.sh">Batch script used in course</a></p>
<p>The following image shows the visualisation of the contigs using Bandage. The contigs are sorted by length.
<img src="file:///workspaces/biol217/resources/bandage_graph.png" alt="Bandage results"></p>
</li>
</ul>
<p>A quick way to count the number of contigs in a fasta file is to use the following code:</p>
<pre class="hljs"><code><div>grep -c &quot;&gt;&quot; final.contigs.fa
</div></code></pre>
<p>Qualty assessment of the contigs is done using <strong><a href="https://quast.sourceforge.net/quast.html">quast</a></strong>. It creates a report that contains information about the contigs. The report is used to decide which parameters to use for binning or reassembly.</p>
<ul>
<li>
<p>Command: <strong><a href="https://quast.sourceforge.net/quast.html">quast</a></strong></p>
</li>
<li>
<p>input: <code>fa</code> (fasta)</p>
</li>
<li>
<p>output: <code>HTML</code> (report)</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/metaquastscript.sh">Batch script used in course</a></p>
</li>
</ul>
<p><strong>Questions to assembly:</strong></p>
<pre><code>- What is your N50 value? Why is this value relevant? 
The N50 value is 2961. This value shows the length of the contigs that are needed to cover half of the total length of the contigs.

- How many contigs are assembled?
56057

- What is the total length of the contigs?
142162752
</code></pre>
<h2 id="4-binning">4. Binning</h2>
<p>The binning is done in the following steps:</p>
<ul>
<li>Mapping of the contigs as preparation for binning.</li>
<li>Binning
<ul>
<li>Refference based</li>
<li>De-Novo by differential coverage</li>
</ul>
</li>
<li>Bin refinement</li>
<li>Bin reassembly</li>
</ul>
<h3 id="preparation-mapping-of-the-contigs">Preparation (mapping of the contigs)</h3>
<p>Binning is done in the <a href="https://anvio.org/">anvi'o</a> environment, which requires a specific file format. To convert the fasta file to the anvi'o file type, the following command is used:</p>
<ul>
<li>
<p>Command: <strong><a href="https://anvio.org/help/main/programs/anvi-script-reformat-fasta/">anvi-script-reformat-fasta</a></strong></p>
</li>
<li>
<p>Input: <code>fa</code> (fasta)</p>
</li>
<li>
<p>Output: <code>fa</code> (fasta anvi'o)</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/anvio_reformat.sh">Batch script used in course</a></p>
</li>
</ul>
<p>Mapping of the contigs is done using <strong><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2</a></strong>. Here the raw reads are mapped onto our assembled contigs.
First the index of the contigs is created:</p>
<ul>
<li>
<p>Command: <strong><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2-build</a></strong></p>
</li>
<li>
<p>Input: <code>fa</code> (fasta anvi'o)</p>
</li>
<li>
<p>Output: <code>bt2</code> (bowtie2 index)</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/bowtie2_build.sh">Batch script used in course</a></p>
</li>
</ul>
<p>Then the raw reads are mapped onto the contigs:</p>
<ul>
<li>
<p>Command: <strong><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2</a></strong></p>
</li>
<li>
<p>Input: <code>bt2</code> (bowtie2 index) + <code>fq</code> (fastq trimmed)</p>
</li>
<li>
<p>Output: <code>sam</code> (sequence alignment map)</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/mappingscript.sh">Batch script used in course</a></p>
</li>
</ul>
<p>The mapping results are converted to a binary alignment map (bam) file:</p>
<ul>
<li>
<p>Command: <strong><a href="http://www.htslib.org/">samtools view</a></strong></p>
</li>
<li>
<p>Input: <code>sam</code> (sequence alignment map)</p>
</li>
<li>
<p>Output: <code>bam</code> (binary alignment map)</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/sam2bam.sh">Batch script used in course</a></p>
</li>
</ul>
<p>The bam file is sorted and indexed:</p>
<ul>
<li>
<p>Command: <strong><a href="https://anvio.org/help/main/programs/anvi-init-bam/">anvi-init-bam</a></strong></p>
</li>
<li>
<p>Input: <code>bam</code> (binary alignment map)</p>
</li>
<li>
<p>Output: <code>bam</code> (sorted) + <code>bam.bai</code> (index)</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/sort_index_bam.sh">Batch script used in course</a></p>
</li>
</ul>
<p>Next we create a contigs database which holds important information about the contigs. The program <strong><a href="https://merenlab.org/2016/06/22/anvio-tutorial-v2/#creating-an-anvio-contigs-database">anvi-gen-contigs-database</a></strong> computes the k-mer frequencies and identifies open reading frames (ORFs) using <a href="https://github.com/hyattpd/Prodigal">Prodigal</a>. On the created contigs database, we perform a search for hidden Markov models (HMMs) using <a href="https://anvio.org/help/7/programs/anvi-run-hmms/">anvi-run-hmms</a>.</p>
<p>This is done using the following command:</p>
<ul>
<li>
<p>Command: <strong><a href="https://merenlab.org/2016/06/22/anvio-tutorial-v2/#creating-an-anvio-contigs-database">anvi-gen-contigs-database</a></strong> + <strong><a href="https://anvio.org/help/7/programs/anvi-run-hmms/">anvi-run-hmms</a></strong></p>
</li>
<li>
<p>Input: <code>fa</code> (fasta anvi'o)</p>
</li>
<li>
<p>Output: <code>db</code> (contigs database)</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/contigdbscript.sh">Batch script used in course</a></p>
</li>
</ul>
<p>Creating anvi'o profiles:</p>
<ul>
<li>
<p>Command: <strong><a href="https://merenlab.org/2016/06/22/anvio-tutorial-v2/#creating-an-anvio-profile">anvi-profile</a></strong></p>
</li>
<li>
<p>Input: <code>bam</code> (sorted) + <code>db</code> (contigs database)</p>
</li>
<li>
<p>Output: <code>db</code> (anvi'o profile)</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/anvio_profile.sh">Batch script used in course</a></p>
</li>
</ul>
<p>Merging the anvi'o profiles:</p>
<ul>
<li>
<p>Command: <strong><a href="https://merenlab.org/2016/06/22/anvio-tutorial-v2/#merging-anvio-profiles">anvi-merge</a></strong></p>
</li>
<li>
<p>Input: <code>db</code> (anvi'o profiles)</p>
</li>
<li>
<p>Output: <code>db</code> (merged profile)</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/anvi_merge.sh">Batch script used in course</a></p>
</li>
</ul>
<h3 id="binning">Binning</h3>
<p>Now the actual binning is done using <a href="https://bitbucket.org/berkeleylab/metabat/src/master/">Metabat2</a> and <a href="https://github.com/BinPro/CONCOCT">CONCOCT</a>, afterwards we choose the best bins by running the <a href="https://github.com/cmks/DAS_Tool">DAS_Tool</a>.</p>
<ul>
<li>Command: <strong><a href="https://anvio.org/help/main/programs/anvi-cluster-contigs/">anvi-cluster-contigs</a></strong></li>
<li>Input: <code>db</code> (merged profile + contigs database)</li>
<li>Output: <code>db</code> (bins database)+ <code>html</code> (interactive report) + <code>fa</code> (fasta)</li>
</ul>
<p>Metabat2 binning:</p>
<ul>
<li>
<p>requires the additon of --driver metabat2 to the anvi-cluster-contigs command</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/metabat2.sh">Batch script used in course</a></p>
</li>
</ul>
<p>CONCOCT binning:</p>
<ul>
<li>
<p>requires the additon of --driver concoct to the anvi-cluster-contigs command</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/concoct.sh">Batch script used in course</a></p>
</li>
</ul>
<p>DAS_Tool for consolidating the bins:</p>
<ul>
<li>
<p>requires the additon of --driver das_tool to the anvi-cluster-contigs command</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/dastool.sh">Batch script used in course</a></p>
</li>
</ul>
<p><strong>Questions to the binning:</strong></p>
<pre><code>Number of Archaea bins you got from MetaBAT2?
3 of 48 bins

Number of Archaea bins you got from CONCOCT?
2 of 156 bins 

Number of Archaea bins you got after consolidating the bins?
2 of 41 bins
</code></pre>
<h2 id="5-mags-quality-assessment">5. MAGs Quality assessment</h2>
<ul>
<li>The quality of the bins is assessed.</li>
<li>The completeness and contamination of the genomes are estimated.
types of contamination:
<ul>
<li>redundant sequences (parts are covered multiple times)</li>
<li>chimeric sequences (contamination by other organisms)</li>
<li>both redundant and chimeric sequences</li>
</ul>
</li>
</ul>
<p>These steps can be done with the generated html file from the binning step (anvi-summarize). Alternatively, we can use Anvi'o's interactive interface.</p>
<p><strong>Questions to the quality assessment:</strong></p>
<pre><code>Which binning strategy gives you the best quality for the Archaea bins??
Metabat2

How many Archaea bins do you get that are of High Quality? 
2 Archaea bins

How many Bacteria bins do you get that are of High Quality?
30 Bacteria bins
</code></pre>
<h2 id="6-bin-refinement">6. Bin refinement</h2>
<ul>
<li>
<p>The bins are refined to remove redundancy and chimeric sequences.</p>
<ul>
<li>Automatic using GUNC</li>
<li>Manual bin refinement	using Anvi'o's interactive interface</li>
</ul>
<p>To detect chimeric bins, we use the program <a href="https://grp-bork.embl-community.io/gunc/">GUNC</a>. The program is run on the fasta file of the Archaea bins. Only the non chimeric bins pass the GUNC filter.</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/gunc.sh">Batch script used in course</a></p>
</li>
</ul>
<p><strong>Questions to the bin refinement:</strong></p>
<pre><code>Do you get Archaea bins that are chimeric? 
Yes, we get chimeric bins from the concoct binning.
A low css score indicates that the bin is not chimeric.

In your own words (2 sentences max), explain what is a chimeric bin.
A chimeric bin is a bin that includes sequences from multiple organisms of different species. 
</code></pre>
<p>Afteer sorting out the chimeric bins wit GUNC, we can use Anvi'o's interactive interface to refine the bins manually.</p>
<p>To access the interactive interface, we use the following commands directly in the terminal:</p>
<pre class="hljs"><code><div>#To request a node
srun --reservation=biol217 --pty --mem=10G --nodes=1 --tasks-per-node=1 --cpus-per-task=1 --nodelist=node### /bin/bash

#To activate the anvio environment
conda activate /home/sunam225/miniconda3/miniconda4.9.2/usr/etc/profile.d/conda.sh/envs/anvio-7.1

#To start the interactive interface with specific command
anvi-refine -c /PATH/TO/contigs.db -C consolidated_bins -p /PATH/TO/merged_profiles/PROFILE.db --bin-id Bin_METABAT__25
</div></code></pre>
<p>In a new terminal window:</p>
<pre class="hljs"><code><div>#Login to the cluster
ssh -L 8060:localhost:8080 sunam###@caucluster-old.rz.uni-kiel.de

#Access the previously requested node
ssh -L 8080:localhost:8080 node###
</div></code></pre>
<p>In google Chrome, we can now access the interactive interface with http://127.0.0.1:8060 .</p>
<p>In the interactive interface, we sort the bins by differerential coverage and remove potential chimeric sequences.</p>
<p><strong>Questions to the bin refinement:</strong></p>
<pre><code>Does the quality of your Archaea improve?
  Completeness before: 97,4% 
  Completeness after: 93,4%
  The redundancy stays the same: 5,3%
  Although the completeness decreased,  the bin is improved because we removed potential chimeric sequences.

How abundant are the Archaea bins in the 3 samples?
  BGR_130708: 0.58
  BGR_130527: 1.14
  BGR_130305: 1.76

Submit your output Figure
</code></pre>
<p><img src="file:///resources/bin_refine_cut.png" alt="Refined bin"></p>
<h2 id="7-taxonomic-classification">7. Taxonomic classification</h2>
<p>Now we add taxonomic annotations to the MAG. This allows us to see present single copy genes and estimate the microbial composition of our metagenome.
In this step we run the following programs in a batch script:</p>
<ul>
<li>
<p>Command: <strong><a href="https://anvio.org/help/main/programs/anvi-run-scg-taxonomy/">anvi-run-scg-taxonomy</a></strong></p>
</li>
<li>
<p>Command: <strong><a href="https://anvio.org/help/main/programs/anvi-estimate-scg-taxonomy/">anvi-estimate-scg-taxonomy</a></strong></p>
</li>
<li>
<p>Command: <strong><a href="https://anvio.org/help/main/programs/anvi-summarize/">anvi-summarize</a></strong></p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/taxonomy.sh">Batch script used in course</a></p>
</li>
</ul>
<p><strong>Questions to the taxonomic classification:</strong></p>
<pre><code>Did you get a species assignment to the Archaea bins previously identified?
  Metabat2: Methanoculleus sp012797575
  CONCOCT: Methanosarcina flavescens

Does the HIGH-QUALITY assignment of the bin need revision?
  Yes, a manual quality check using blast. And further refinement of the bins is necessary.
</code></pre>
<h1 id="pangenomics">Pangenomics</h1>
<p>The aim of pangenomics is to compare the assembled MAGs to reference genomes to gain insight into the underlying phylogenetics and functional potential of the MAGs.</p>
<h2 id="pangenoimcs-workflow">Pangenoimcs workflow</h2>
<ol>
<li>Evaluate the contig databases</li>
<li>Create a pangenome database</li>
<li>Compare the databases using ANI</li>
<li>Visualize the results</li>
<li>Interpretation of the results</li>
</ol>
<h2 id="1-evaluate-the-contig-databases">1. Evaluate the contig databases</h2>
<p>For evaluation if the contig databases we use Anvio interactive interface. We can use the following commands to access the interface:</p>
<pre class="hljs"><code><div>#get direct access to a HPC compute node
srun --reservation=biol217 --pty --mem=10G --nodes=1 --tasks-per-node=1 --cpus-per-task=1 /bin/bash

#activate the conda environment
conda activate /home/sunam225/miniconda3/miniconda4.9.2/usr/etc/profile.d/conda.sh/envs/anvio-7.1

#move to working directory
cd /path/to/

# start anvi'o interactive display(-P set the port)
anvi-display-pan -g *-GENOMES.db -p *-PAN.db -P 8089

# start tunnel in new Terminal(adjust port from pevious step &amp; node)
ssh -L 8060:localhost:8089 sunam230@caucluster-old.rz.uni-kiel.de
ssh -L 8089:localhost:8089 node010

#open with link in browser:
http://127.0.0.1:8060
</div></code></pre>
<p>Task: Take some time to click through the views and compare the MAGs. Add a screenshot of your output to your documentation. Answer the following Questions:</p>
<p>Question: How do the MAGs compare in size and number of contigs to the full genome?</p>
<p>Answer: The MAGs are much shorter. For example N50 shows the average length of fragments in different Bins. Here &quot;Methano_Bin13: 22,585&quot; is much better than &quot;Methano_Bin5: 4,826&quot;</p>
<table>
<thead>
<tr>
<th>Contigs Stats</th>
<th>Methano_Bin10</th>
<th>Methano_Bin13</th>
<th>Methano_Bin1</th>
<th>Methano_Bin3</th>
<th>Methano_Bin5</th>
<th>Methano_Bin8</th>
<th>Methano_Bin9</th>
<th>Methano_Mflavescens</th>
</tr>
</thead>
<tbody>
<tr>
<td>N50</td>
<td>4,894</td>
<td>22,585</td>
<td>5,434</td>
<td>13,114</td>
<td>4,826</td>
<td>19,343</td>
<td>13,691</td>
<td>3,283,688</td>
</tr>
</tbody>
</table>
<p>Question: Based on the contig numbers, sizes and number of marker genes (HMM hits), which two MAGs are the best and which is the worst?</p>
<p>Answer:</p>
<table>
<thead>
<tr>
<th>MAGs</th>
<th>Number of contigs</th>
<th>Size of contigs</th>
<th>Number of marker genes</th>
</tr>
</thead>
<tbody>
<tr>
<td>Best</td>
<td>Methano_Bin13</td>
<td>Methano_Bin13</td>
<td>Methano_Bin13</td>
</tr>
<tr>
<td>Worst</td>
<td>Methano_Bin5</td>
<td>Methano_Bin10</td>
<td>Methano_Bin5</td>
</tr>
</tbody>
</table>
<p><img src="file:///resources/Contigstats1.png" alt="Contig stats">
<img src="file:///resources/Contigstats3.png" alt="Contig stats">
<img src="file:///resources/Contigstats2.png" alt="Contig stats"></p>
<h2 id="2-create-a-pangenome-database">2. Create a pangenome database</h2>
<p>In this step we create a pangenome database from the MAGs. We use the following commands:</p>
<h3 id="create-an-external-genomes-file">Create an external genomes file:</h3>
<pre class="hljs"><code><div>anvi-script-gen-genomes-file --input-dir ? -o external-genomes.txt
</div></code></pre>
<p>The created file external-genomes.txt can be accessed in the terminal:</p>
<pre class="hljs"><code><div>external-genomes.txt

name	contigs_db_path
Methano_Bin1	/work_beegfs/sunam230/Day6/02_contigs-dbs/Bin1.db
Methano_Bin10	/work_beegfs/sunam230/Day6/02_contigs-dbs/Bin10.db
Methano_Bin13	/work_beegfs/sunam230/Day6/02_contigs-dbs/Bin13.db
Methano_Bin3	/work_beegfs/sunam230/Day6/02_contigs-dbs/Bin3.db
Methano_Bin5	/work_beegfs/sunam230/Day6/02_contigs-dbs/Bin5.db
Methano_Bin8	/work_beegfs/sunam230/Day6/02_contigs-dbs/Bin8.db
Methano_Bin9	/work_beegfs/sunam230/Day6/02_contigs-dbs/Bin9.db
Methano_Mflavescens	/work_beegfs/sunam230/Day6/02_contigs-dbs/Mflavescens.db
</div></code></pre>
<h3 id="estimate-genome-completeness">Estimate genome completeness</h3>
<pre class="hljs"><code><div>anvi-estimate-genome-completeness -e external-genomes.txt &gt; genome-completeness.txt
</div></code></pre>
<p>Question: How do the bins compare to isolate genomes? Would you remove one, based on the output of the completeness estimation?</p>
<p>ANSWER:
Bin10 &amp; Bin5 are below 70% completeness so we remove them.
external-genomes-final.txt</p>
<pre class="hljs"><code><div>╒═════════════════════╤══════════╤══════════════╤════════════════╤════════════════╤══════════════╤════════════════╕
│ genome name         │ domain   │   confidence │   % completion │   % redundancy │   num_splits │   total length │
╞═════════════════════╪══════════╪══════════════╪════════════════╪════════════════╪══════════════╪════════════════╡
│ Methano_Bin1        │ ARCHAEA  │          0.8 │          86.84 │           5.26 │          275 │        1375677 │
├─────────────────────┼──────────┼──────────────┼────────────────┼────────────────┼──────────────┼────────────────┤
│ Methano_Bin10       │ ARCHAEA  │          0.3 │          56.58 │              0 │          286 │        1336430 │
├─────────────────────┼──────────┼──────────────┼────────────────┼────────────────┼──────────────┼────────────────┤
│ Methano_Bin13       │ ARCHAEA  │            1 │            100 │           5.26 │          182 │        2194666 │
├─────────────────────┼──────────┼──────────────┼────────────────┼────────────────┼──────────────┼────────────────┤
│ Methano_Bin3        │ ARCHAEA  │          0.9 │          96.05 │           3.95 │          180 │        1769119 │
├─────────────────────┼──────────┼──────────────┼────────────────┼────────────────┼──────────────┼────────────────┤
│ Methano_Bin5        │ ARCHAEA  │          0.3 │          43.42 │              0 │          334 │        1582595 │
├─────────────────────┼──────────┼──────────────┼────────────────┼────────────────┼──────────────┼────────────────┤
│ Methano_Bin8        │ ARCHAEA  │          0.9 │          93.42 │           6.58 │          145 │        1795577 │
├─────────────────────┼──────────┼──────────────┼────────────────┼────────────────┼──────────────┼────────────────┤
│ Methano_Bin9        │ ARCHAEA  │          0.8 │          78.95 │           2.63 │          255 │        2637263 │
├─────────────────────┼──────────┼──────────────┼────────────────┼────────────────┼──────────────┼────────────────┤
│ Methano_Mflavescens │ ARCHAEA  │            1 │          94.74 │              0 │          164 │        3283688 │
╘═════════════════════╧══════════╧══════════════╧════════════════╧════════════════╧══════════════╧════════════════╛
</div></code></pre>
<h3 id="remove-unwanted-genomes">Remove unwanted genomes</h3>
<p>We remove the genomes with low completeness (&lt;70%) and high redundancy (&gt;10%) from the contigs directory and re-create the external-genomes file.</p>
<pre class="hljs"><code><div>anvi-script-gen-genomes-file --input-dir ? -o external-genomes-final.txt
</div></code></pre>
<h3 id="create-a-pangenome-database">Create a pangenome database</h3>
<p>For the pangenome analysis we need to create two databases:</p>
<ul>
<li>
<p><a href="https://anvio.org/help/7/artifacts/genomes-storage-db/">genomes-storage.db</a> which contains:</p>
<ul>
<li>all genome fasta files</li>
<li>the gene annotations (HMMs, SCGs)</li>
<li>genome comparisons</li>
</ul>
</li>
<li>
<p><a href="https://anvio.org/help/main/programs/anvi-pan-genome/">pan-genome.db</a> which contains:</p>
<ul>
<li>genome similarities based on gene amino acid sequences</li>
<li>resolved gene clusters</li>
<li>any post-analysis of gene clusters, downstream analyses and visualisations</li>
</ul>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/anvi_genomes.sh">Batch script used in course</a></p>
</li>
</ul>
<h2 id="3-compare-the-databases-using-ani">3. Compare the databases using ANI</h2>
<p>In this step we use the <a href="https://anvio.org/help/main/programs/anvi-compute-genome-similarity/">anvi-compute-genome-similarity</a> command to compare the genomes in the genomes-storage.db database. The resulting file shows the similarity between all genomes in the database.</p>
<pre><code>[Batch script used in course](https://github.com/YanGiencke/biol217/blob/main/scripts/similarity.sh)
</code></pre>
<h2 id="4-visualize-the-results">4. Visualize the results</h2>
<p>For visualisation we use the interactive interface of anvi'o. It is stated a described above, however we use the command:</p>
<pre class="hljs"><code><div>anvi-display-pan -p Biol217/Biol217-PAN.db -g Methanogen-GENOMES.db -P 8089
</div></code></pre>
<p>Task: Scroll to the top of the help and find out which INPUT FILES you need. Write the command and use the additional flag -P. What is the -P flag for?</p>
<p>Answer:<br>
INPUT: -p Pan-Genome.db -g Combined-Genomes.de
-P sets the port example: 8089</p>
<h2 id="5-interpretation-of-the-results">5. Interpretation of the results</h2>
<p>The interactive interface of anvi'o is very powerful and allows you to explore the results of your analysis. We will use it to explore the results of the pangenome analysis.</p>
<p>Anvio interactive Tasks:</p>
<h3 id="genome-similarity">Genome similarity</h3>
<ul>
<li>Remove combined homogeneity, functional homogeneity, geometric homogeneity, max num parsimonay, number of genes in gene cluster and number of genomes gene cluster has hits from the active view. Tip: Play with Height</li>
<li>Create a &quot;Bin-highlight&quot; including alls SCGs and name it accordingly. How to?</li>
</ul>
<p><img src="file:///workspaces/biol217/resources/1interactive.png" alt="1 &amp; 2"></p>
<ul>
<li>
<p>Cluster the genomes based on Frequency
<img src="file:///workspaces/biol217/resources/3interactive.png" alt="3"></p>
</li>
<li>
<p>Question:</p>
<ul>
<li>Based on the frequency clustering of genes, do you think all genomes are related? Why?</li>
<li>Highlight your reference genome in one color, its closest relative in a similar one, and distict genomes in a third colour.</li>
<li>How does the reference genome compare to its closest bin? Tip: Consider the genome depiction and layers above</li>
</ul>
</li>
</ul>
<p>Answer: No. Only Bin 9 is closely related to the reference Methano Mflavescens. The other Bins are related to each other.</p>
<p><img src="file:///workspaces/biol217/resources/4interactive.png" alt="4"></p>
<ul>
<li>
<p>Go to Layers and remove Num gene clusters, Singeltons, Genes per kbp and Total length from view. Add ANI_percentage_identity to the view and play with the threshold.</p>
</li>
<li>
<p>Questions: What ranges are used determine a prokaryotic species? How high can you go until you see changes in ANI? What does the ANI clustering tell you about genome relatedness?</p>
<p>Answer:</p>
<ul>
<li>over 95% identity for DNA from one species</li>
<li>differences between the two clusters appear upwards from 98%</li>
<li>The higher the saturation of a tupel, the higher the relatedness of the two Bins</li>
</ul>
</li>
</ul>
<p><img src="file:///workspaces/biol217/resources/5.png" alt="5"></p>
<h3 id="tasks-functional-profiling">TASKS: Functional Profiling</h3>
<ul>
<li>
<p>Using the Search Function, highlight all genes in the KEGG Module for Methanogenesis</p>
<ul>
<li>Create a new bin called &quot;Methanogenesis&quot; and store your search results in this bin.</li>
<li>Question: How are Methanogenesis genes distributed across the genome?</li>
</ul>
<p>Answer: Through the whole Pangenome.</p>
</li>
</ul>
<p><img src="file:///workspaces/biol217/resources/6.png" alt="6"></p>
<ul>
<li>
<p>Google COG Categories and select one you are interesed in. Create a new bin, find your Category in the Pangenome and add it to this selection.</p>
<p>T - COG0642 - Signal transduction histidine kinase
<img src="file:///workspaces/biol217/resources/7.png" alt="7"></p>
</li>
<li>
<p>Save your state and export this view as .svg</p>
</li>
</ul>
<p><a href="resources/pangenome.svg">pangenome.svg</a></p>
<h3 id="tasks-functionalgeometric-homogeneity-and-their-uses">TASKS: Functional/geometric homogeneity and their uses</h3>
<ul>
<li>
<p>Using search parameters, find a gene which occurs:</p>
<ul>
<li>in all genomes</li>
<li>a maximum of 1 times (Single copy gene)</li>
<li>has a high variability in its functional homogeneity (max. 0.80)</li>
</ul>
<p>This gene will be highly conserved, but has diversified in its AA make-up.</p>
<p>Highlight the found genes on the interface. Inspect one of the gene-clusters more closely (Inspect gene-cluster).
<img src="file:///workspaces/biol217/resources/8.png" alt="8"></p>
</li>
</ul>
<p>Question: What observations can you make regarding the geometric homogeneity between all genomes and the functional homogeneity?</p>
<p>The AA sequences start later in some Bins these are not geometrically homogenous
<img src="file:///workspaces/biol217/resources/9.png" alt="9"></p>
<h3 id="bonus-blastkoala">BONUS: BlastKoala</h3>
<p>Outside of anvi'o there are a range of tools available to investigate your organisms metabolism. One of these is BlastKOALA, which generates a metabolic profile of your genome based on the KEGG database.</p>
<ul>
<li>
<p>Task: Check out the BlastKOALA Results for this Methanogen.
<img src="file:///workspaces/biol217/resources/10.png" alt="10"></p>
</li>
<li>
<p>Reconstruct its pathways and check out what it can do.</p>
</li>
<li>
<p>Question: Can the organism do methanogenesis?</p>
</li>
</ul>
<p>Answer: Yes, the methanogenesis pathway shows up in Blast Koala.
<img src="file:///workspaces/biol217/resources/11.png" alt="11"></p>
<ul>
<li>Does it have genes similar to a bacterial secretion system?
Yes some genes for the bacterial sercretion system are present.
<img src="file:///workspaces/biol217/resources/12.png" alt="12"></li>
</ul>
<h1 id="rna-seq">RNA-Seq</h1>
<h2 id="central-dogma">Central dogma</h2>
<p>DNA -&gt; RNA -&gt; Protein</p>
<ul>
<li>
<p>DNA: deoxyribonucleic acid (genetic material)</p>
</li>
<li>
<p>RNA: ribonucleic acid (messenger)</p>
</li>
<li>
<p>Protein: amino acid chain (functional)</p>
</li>
<li>
<p>optimal read length: 50 bp single end, 100 bp paired end</p>
</li>
<li>
<p>optimal replicate number: 3-5</p>
</li>
</ul>
<h2 id="rna-seq-workflow">RNA-Seq workflow</h2>
<ol>
<li>Extract RNA</li>
<li>Produce cDNA</li>
<li>Sequencing</li>
<li>Bioinformatics</li>
<li>Analysis</li>
</ol>
<h2 id="1-extract-rna">1. Extract RNA</h2>
<ul>
<li>RNA is extracted from cells</li>
</ul>
<h2 id="2-produce-cdna">2. Produce cDNA</h2>
<ul>
<li>cDNA is produced from RNA using reverse transcription</li>
</ul>
<h2 id="3-illumina-sequencing">3. Illumina sequencing</h2>
<p>Illumina sequencing is a method to sequence DNA or RNA. It is based on the principle of sequencing by synthesis. The DNA or RNA(cDNA) is fragmented into small pieces and the fragments are attached to a flow cell. The flow cell is then placed in a sequencing machine. The sequencing machine uses a laser to read the DNA fragments by recording the fluorescent signal of the attached nucleotides. The sequencing machine can read millions of DNA fragments in parallel.</p>
<h2 id="alternatively-download-sequenced-data-from-papers">Alternatively download sequenced data from papers</h2>
<ul>
<li>Go to paper</li>
<li>Search for key words (SRA, SRR, accession number)</li>
<li>Follow link to NCBI SRA (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE85456)</li>
<li>Navigate to SRA and extract SRR numbers of interest</li>
<li>Import the data into your computer with the following command:</li>
</ul>
<pre class="hljs"><code><div>#Samples 
fasterq-dump SRR***, SRR***, SRR***, SRR***
#Metadata
grabseqs sra -l -t 4 -m metadata.csv SRP081251
</div></code></pre>
<ul>
<li>SRR numbers
<ul>
<li>SRR4018517</li>
<li>SRR4018516</li>
<li>SRR4018515</li>
<li>SRR4018514</li>
</ul>
</li>
</ul>
<h2 id="4-bioinformatics">4. Bioinformatics</h2>
<p>The bioinformatics workflow is similar to the sequencing pipeline described above. However, in this case we automate the process by using a program called READemption. READemption is a pipeline for the analysis of RNA-Seq data, it is a wrapper for several other programs.</p>
<p>Before we start the READemption pipeline, we might have to perform quality control and clean up if the data is unprocessed. For this we use <strong><a href="https://home.cc.umanitoba.ca/~psgendb/doc/fastqc.help">fastqc</a></strong> and <strong><a href="https://github.com/OpenGene/fastp#fastp">fastp</a></strong>. Then we can create the folder structure neccessary for READemption and move the data into the correct folders. We use the following command to create the folder structure:</p>
<pre class="hljs"><code><div>reademption create --project_path READemption_analysis --species salmonella=&quot;Salmonella Typhimurium&quot;
</div></code></pre>
<p>Lastly we need to make sure that the data has the correct format and headers.
Ater that all steps are executed in one batch script by READemption.</p>
<p><a href="https://github.com/YanGiencke/biol217/blob/main/scripts/READemption.sh">Batch script used in course</a></p>
<h1 id="introduction-to-r">Introduction to R</h1>
<p>R is a programming language and software environment for statistical computing and graphics. In a short introduction we learned the basics of R and how to use it for data visualization.
We used the following code:</p>
<pre class="hljs"><code><div># Title: Introduction to R

# To print a value, use the print() function.
print(&quot;Hello, world!&quot;)

# To print a value without the print() function, use the value itself.

# to show the current working directory, use the getwd() function.
getwd()

# To create a new directory, use the dir.create() function.
# dir.create(&quot;data&quot;)
# dir.create(&quot;data/raw_data&quot;)

# R has 5 basic data types: logical, integer, numeric, complex, and character.
# To assign a value to a variable, use the assignment operator, &lt;-.
x &lt;- 1:10
t &lt;- TRUE

# To create a sequence of numbers, use the colon operator, :.
z &lt;- 1:10

# To create a sequence of numbers with a step size, use the seq() function.
w &lt;- seq(1, 10, by = 2)

# Data types
# To check the data type of a variable, use the class() function.
class(x) # numeric
class(t) # logical

# To check the length of a variable, use the length() function.
length(x)

# R has 3 basic data structures: vectors, matrices, and data frames.
# To create a vector, use the c() function.
y &lt;- c(1, 2, 3, 4, 5)

# To create a matrix, use the matrix() function.
m &lt;- matrix(1:9, nrow = 3, ncol = 3)

# To create a data frame, use the data.frame() function.
df &lt;- data.frame(x = 1:10, y = 1:10)

# R operators
# Mathematical operators: +, -, *, /, ^, %%
# Logical operators: &gt;, &lt;, &gt;=, &lt;=, ==, !=
# Logical operators: &amp;, |, !


# Test your knowledge
# using iris dataset
data(&quot;iris&quot;)
View(iris)

plot(iris)

boxplot(iris$Sepal.Length ~ iris$Species)

setwd(&quot;/home/kurs/Desktop/biol217/R/intro&quot;)


# install.packages(&quot;ggplot2&quot;)
# install.packages(c(&quot;readxl&quot;, &quot;plotly&quot;))
install.packages(&quot;tidyverse&quot;)
install.packages(&quot;svglite&quot;, dependencies = TRUE)
library(ggplot2)
library(svglite)

setRepositories()

# Plotting with ggplot2
ggplot(iris, aes(x = Sepal.Length, y = Sepal.Width, color = Species, size = Petal.Length*Petal.Width)) +
  geom_point()

plot1 &lt;- ggplot(iris, aes(x = Sepal.Length, y = Sepal.Width, color = Species, size = Petal.Length*Petal.Width)) +
  geom_point()

# Saving plots as different files
ggsave(&quot;Iris.tiff&quot;, plot1, width = 20, height = 10, unit = &quot;cm&quot;, dpi = 300)

ggsave(&quot;Iris_compressed.tiff&quot;, plot1, width = 20, height = 10, unit = &quot;cm&quot;, dpi = 300, compression = &quot;lzw&quot;)

ggsave(&quot;Iris2.pdf&quot;, plot1)

ggsave(&quot;Iris2.svg&quot;, plot1)

svglite(&quot;Iris.svg&quot;, width = 20, height = 10, pointsize = 12)



# Reshaping data with tidyr
library(tidyr)
library(plotly)
library(readxl)


# spread iris dataset to wide format using tidyr spread() function
new_table &lt;- spread(data = iris, key = iris$Species, value = c(iris$Sepal.Length, iris$Sepal.Width, iris$Petal.Length, iris$Petal.Width))



# visualizing distribution
hist(iris$Sepal.Length)
hist(iris$Sepal.Width)
hist(iris$Petal.Length)
hist(iris$Petal.Width)

# visualizing outliers
boxplot(iris$Sepal.Length)
boxplot(iris$Sepal.Width)
boxplot(iris$Petal.Length)
boxplot(iris$Petal.Width)

# Statistical analysis of normal distributed data
# 1 sample t-test
# 2 sample t-test
# more than 2 samples ANOVA



# More plotting with ggplot2
data()
View(DNase)

ggplot(DNase, aes(x = conc, y = density)) + geom_point()
 + xlab(&quot;Concentration&quot;) + ylab(&quot;Density&quot;) + ggtitle(&quot;DNase I activity&quot;)
  + ggsave(&quot;DNase.jpg&quot;)

ggplot(DNase, aes(x = conc, y = density)) + geom_point() + ggsave(&quot;DNase.jpg&quot;)
hist(DNase$density)
hist(DNase$conc)
</div></code></pre>
<p>The following plots were created:
<img src="file:///R/intro/plot.png" alt="irisall"></p>
<p><img src="file:///R/intro/Iris.png" alt="iris"></p>
<p><img src="file:///R/intro/boxplot.png" alt="iris"></p>
<p><img src="file:///R/intro/DNase.jpg" alt="iris"></p>
<h1 id="ribo-seq">RIBO-Seq</h1>
<h2 id="what-is-ribo-seq">What is Ribo-Seq?</h2>
<p>Central dogma of molecular biology: DNA -&gt; RNA -&gt; <strong>Translatome</strong> -&gt; Protein
The Translatome is the mRNA bound to ribosomes.
Simillar to RNA-Seq the mRNA is sheared but only the regions with ribosomes attached are sequenced. This is called Ribo-Seq.
Since the ribosomes only bind to the coding regions of the mRNA, the Ribo-Seq data can be used to identify the genes that are actually being expressed.
In combination with RNA-Seq data, the Ribo-Seq data can be used to identify the genes that are being translated.</p>
<p>Ribosome protected fragments (RPFs) are about 25–40 bp in length.</p>
<p>In the course we used data fron the paper: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6347204/pdf/pone.0211430.pdf">Role of CsrA in stress responses and metabolism important for Salmonella virulence revealed by integrated transcriptomics</a>. The image below shows the experimental design from the paper.
<img src="file:///workspaces/biol217/resources/Experimental_design.png" alt="Ribo-Seq"></p>
<h2 id="integrated-genome-browser-igb">Integrated Genome Browser (IGB)</h2>
<p>The Integrated Genome Browser (IGB) is a program used for analysing differential gene expression based on RNA-Seq and RIBO-Seq data. To use the program we need to load the fasta files for the genome and the annotation file (.gff). Then we can load the coverage files (.bw).
The following picture shows the differential expression of the csrA gene in the wildtype and Mutant strain described in the paper.
<img src="file:///resources/Expression.png" alt="csrA">
The csrA gene starts at base pair 2991371 and ends at 2991185.In Integerated Genome Browser we can visualize the differential expression of genes between the two conditions.</p>
<p>Information about the csrA gene:</p>
<ol>
<li>Start Codon: ATG</li>
<li>Stop Codon: TAA</li>
<li>Length in Amino Acids: 61 (does not include stop codon)</li>
<li>Shine-Dalgarno sequence (consensus AGGAGG at -7 to -4): AGGAG at -7</li>
<li>Upsream gene: alaS (STM14_3413)</li>
<li>Is the csrA gen translated? Yes, because the RIBO-Seq data covers the whole gene.</li>
</ol>
<h2 id="r-code-for-ribo-seq-analysis">R code for RIBO-Seq analysis</h2>
<p>Using R we can identify more differentially expressed genes and visualize the expression in a heatmap. The following code was used create the heatmap.</p>
<pre class="hljs"><code><div>library(readxl)
library(ggplot2)

sheet1 &lt;- read_excel(&quot;csrA-WT_sorted.xlsx&quot;, sheet = &quot;Sheet4&quot;)
plot1 &lt;- ggplot(sheet1, aes(Seq_type, Name, fill=log2fold_change))+geom_tile() +
  ylab(&quot;Genes&quot;) +xlab(&quot;Sequencing Type&quot;)+ labs(fill=&quot;Change in log2&quot;) + ggtitle(&quot;Differential gene expression&quot;) + 
  geom_text(aes(label=Name))+theme(axis.text.y = element_blank()) + ggsave(&quot;Differential_gene_expression1.png&quot;)

plot1

ggsave(&quot;Differential_gene_expression1&quot;)
</div></code></pre>
<p><img src="file:///R/intro/Differential_gene_expression1.png" alt="plot1"></p>
<pre class="hljs"><code><div>sheet2 &lt;- read_excel(&quot;csrA-WT_sorted.xlsx&quot;, sheet = &quot;Sheet3&quot;)
plot2 &lt;- ggplot(sheet2, aes(Name, log2FC_TE_final))+geom_bar(stat = &quot;identity&quot;) +
  xlab(&quot;Genes&quot;) +ylab(&quot;Change in log2&quot;)+ ggtitle(&quot;Differential gene expression&quot;) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
plot2

ggsave(&quot;Differential_gene_expression2.png&quot;)
</div></code></pre>
<p><img src="file:///R/intro/Differential_gene_expression2.png" alt="plot2"></p>
<pre class="hljs"><code><div>sheet3 &lt;- read_excel(&quot;csrA-WT_sorted.xlsx&quot;, sheet = &quot;Sheet5&quot;)
plot3 &lt;- ggplot(sheet3, aes(Seq_type, Identifier, fill=log2FC))+geom_tile()+
  ggtitle(&quot;Differential gene expression&quot;)+ylab(&quot;Genes&quot;) +xlab(&quot;Sequencing Type&quot;)+ 
  labs(fill=&quot;Change in log2&quot;)+theme(axis.text.y = element_blank())
plot3

ggsave(&quot;Differential_gene_expression3.png&quot;)
</div></code></pre>
<p><img src="file:///R/intro/Differential_gene_expression3.png" alt="plot3"></p>

</body>
</html>
